### ISSUE TITLE
The training gets killed

### ISSUE THREAD

**Issue by Mshz2** [jacobgil/pytorch-grad-cam#117]
Hi guys, thanks for your nice repo. I am using your code to observe the model during training loop. However, after some epochs the python run is suddenly killed. I am using pytorch for my training and validation. I also looked up to my memory by 'htop' command in another terminal and noticed that my memory usage is increasing exponentially with the epochs iterations. And after just 11 epochs, 441GB/441GB of memory is already used.
here is some of my implementation:

```
def cam_extractor(network, input_tensor, rgb_img):

    target_layer = network.lastact[-1]
    cam = GradCAMPlusPlus(model=network, target_layer=target_layer, use_cuda=True)
    # If None, returns the map for the highest scoring category.
    # Otherwise, targets the requested category.
    target_category = 3
    grayscale_cam = cam(input_tensor=input_tensor,
                        target_category=target_category,
                        aug_smooth=False,
                        eigen_smooth=False)

    grayscale_cam = grayscale_cam[0, :]
    cam_image = show_cam_on_image(rgb_img, grayscale_cam, use_rgb=True)

    cam_image = cv2.cvtColor(cam_image, cv2.COLOR_RGB2BGR)
    cv2.imwrite(f'{GradCAMPlusPlus}_cam.jpg', cam_image)
    return cam_image
```
and the training loop is sth below. The Memory consumption is stable if I do not use cam_extractor function which is defined above.
```
for epoch in range(x):
        network.train()
        for batch in train_load: # Get Batch
               images, labels = batch 
               images, labels = images.to(device), labels.to(device)
              .......
       network.eval()
       output_cam = cam_extractor(network, input_tensor, rgb_img, counter, counter_cam)  #I wanna have cam for each epoch

      with torch.no_grad():
          for batch_valid in vali_load: #Get Validation Batch
                   valid_images, valid_labels = batch_valid
                   valid_images, valid_labels = valid_images.to(device), valid_labels.to(device)

```

The error I get is just below:
`
epoch 10 total_correct (train): 42386 total_correct (validation): 7730 loss (train): 90.15721723437309 loss (validation): 27.10268247127533
./train.sh: line 6: 20730 Killed                  python train.py --GPU '0' --dataset cifar10 --data_loc './datasets/cifar10'
(torchpy37) mypc@mypc:~/notebooks/CAM$ ./train.sh 1
`

the train.sh is just a bach file for calling the python file containing: python train.py --GPU '0' --dataset cifar10 --data_loc './datasets/cifar10'

**Comment by jacobgil** [comment]
Hi, sorry for the late reply.
I'm going to look into this in the next few days.

Can you please try defining the cam `cam = GradCAMPlusPlus(model=network, target_layer=target_layer, use_cuda=True)` once outside of the loop, instead of re-creating it for every image?

**Comment by jacobgil** [comment]
I'm pretty sure this line:
       output_cam = cam_extractor(network, input_tensor, rgb_img, counter, counter_cam)  #I wanna have cam for each epoch
Is causing the leak.
You can create it once outside the training loop, and it will use the updating model.
Closing this issue, please re-open if there is still a leak after doing this.
